### DISCUSSION TITLE
Extension for intellisense for Latex in Markdown, problem with embedded-content

### DISCUSSION THREAD

**Discussion by StefanoDeVuono** [microsoft/vscode-discussions#1344] [Extension Development QnA] [answered]
**Summary**
Hi, I'm developping an extension to allow intellisense support for Latex in Markdown as it would greatly increase my speed and ability to take notes. Hopefully, it could be expanded to other languages eventually.

The design I'm trying right now is to capture when I'm in a math section and create a virtual document to run commands against.

I'm basing it off of both VS code's [embedded language request forwarding example](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-request-forwarding) and [this archived project to allow intellisense in Markdown](https://github.com/tamuratak/vscode-markdown-intellisense/).

When using `embedded-content` as per VS Code's example repo, I don't get the right code completions. Using an actual temp file written to disk, I'm able to get the correct ones.

**Question**
Is there a way to send a virtual document to `vscode.executeCompletionItemProvider` and have it *consistently* return the correct completions?

**Details**
In my [`CompletionProvider.ts`](https://github.com/StefanoDeVuono/math-intellisense-markdown/blob/main/src/CompletionProvider.ts) file, I have code that can *either*:
  - use a virtual document with `embedded-content` as per VS Code's example repo
    **or**
  - Use an actual temp file written to disk
  
  Here are the lines in question:
  ```ts
  /**  START - Using Temporary Document */
// const virtualDocument = await createTempDocument(document, node)
// const virtualPosition = createTempPosition(position, node)
// const vdocUri = virtualDocument?.uri
/** END -  Using Temporary Document */

/** START - Using Virtual Document */
this.virtualDocument.update!(document, node, position)

const vdocUri = this.virtualDocument.uri
const virtualPosition = this.virtualDocument.position
/** END - Using Virtual Document */
  ```
  
  Then, we execute the command:
  ```ts
  itemList = await commands.executeCommand<CompletionList>(
  'vscode.executeCompletionItemProvider',
  vdocUri,
  virtualPosition,
  context.triggerCharacter
)
```

**Videos and additional info**
Here's the output using a virtual document:
https://github.com/user-attachments/assets/fc994847-7edf-4d6a-b635-1d2cca8abcac

And here's the output using a temporary document:
https://github.com/user-attachments/assets/bad9ba8f-966f-469d-9d34-59da3f111f66

With the virtual document, the completion items from the `\` character are what you'd see in a text document (although the items from the `@` character are correct).

Using a temp file, the completion items from the `\` character are correct (as well as those from `@`).

It seems like the `vscode.executeCompletionItemProvider` command is reacting differently to each. The videos contain outputs of the variables `vdocUri` and `virtualPosition`. 

Additionally, as you can see in the videos above, the virtual document and temp file end up storing the same string.

**Comment by isidorn** [comment]
I do not know. Maybe @mjbvz might be able to help.

**Comment by mjbvz** [comment]
Have you tried debugging the latex extension to see if it's the root cause? It's possible that it returns different completions for files on disk (even temp ones) compared to those for virtual files

**Reply by StefanoDeVuono** [reply]
Thanks for the suggestions!

I've tried debugging the [LaTeX-Workshop](LaTeX-Workshop) extension directly.
 - When I cloned LaTeX-Workshop and started debugging it on a latex file, every time I typed `\`, `macro.js`'s `provide()` function gave suggestions.
 - Working in my extension I put breakpoints in the LaTeX-Workshop in VS Code's extension directory (note this version of the extension is compiled) and only reached them on startup. I didn't reach them after that. Including when triggering completions. The breakpoints were in the `Provider#provide()`, `Provider#completion`, `AtProvider#provideCompletionItems()`, `AtProvider#provide()` as well as in `macro.js`'s `provide()` function.

It's a bit challenging to debug my extension once it passes off control to other extensions or to the Extension Host (where a lot of stuff seems to be happening in a minified file). Perhaps the `vscode.executeCompletionItemProvider` command is sent directly to some to somewhere else that can give these suggestions?

> It's possible that it returns different completions for files

I did find something along those lines when *not* debugging. When I start up my extension, for a quick second, the `\` key will trigger latex suggestions but only *sometimes* and only the first time running it. I have to run it again and then only sometimes it works. Here's a quick video. You'll notice the markdown extensions flash in the suggestions menu for a sec before being replaced by the LaTeX ones. However, this list of completion items may be incomplete.

https://github.com/user-attachments/assets/4d47252a-e9c5-49c9-8c03-0f0c30d7393c

**Reply by StefanoDeVuono** [reply]
Also, is there a way to debug the Extension Host? Is there a better way to debug a secondary extension (other than just using the transpiled JS in the local VS Code extensions directory)?

**Reply by mjbvz** [reply]
You can easily debug the extension host if you build and run VS Code out of sources. However this may be very confusing if the the Latex extension code is minified

Instead it may be better to try debugging the latex extension out of its source to see if you can figure out what's going on there

**Reply by robconery** [reply]
Hi @StefanoDeVuono did this answer help? Would love to close if so.

**Reply by StefanoDeVuono** [reply, marked as answer]
@mjbvz, @isidorn, and @robconery,

Thank you for all your help.

I was able to debug existing extensions by starting the extension debugger process out of their directories. For example, to debug LaTeX-Workshop:
```sh
cd .vscode/extensions/james-yu.latex-workshop-10.2.1/
code .
```

Then, run the debugger, setting up a configuration like so if necessary:
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Run Extension",
      "type": "extensionHost",
      "request": "launch",
      "args": [
        "--extensionDevelopmentPath=${workspaceFolder}"
      ],
      "outFiles": [
        "${workspaceFolder}/out/**/*.js"
      ],
      "preLaunchTask": "${defaultBuildTask}"
    }
  ]
}
```

Now, I have an [extension in the marketplace](https://marketplace.visualstudio.com/items?itemName=StefanoDeVuono.markdown-math-intellisense) if others want to use it or [expand on it](https://github.com/StefanoDeVuono/math-intellisense-markdown).
